File: test_runner/test_runner.go
package main

import (
	"fmt"
	"os"
)

type testRunner struct {
	failedTests []string
}

func (runner *testRunner) runTest(test func(), name string) {
	defer func() {
		if err := recover(); err != nil {
			runner.failedTests = append(runner.failedTests, fmt.Sprintf("%s: %v", name, err))
			fmt.Print(" fail\n")
		}
	}()

	fmt.Printf("%s...", name)
	test()
	fmt.Print(" pass\n")
}

func main() {
	runner := &testRunner{}
	/* replace: tests */
	for _, failure := range runner.failedTests {
		fmt.Printf("%s\n", failure)
	}
	if len(runner.failedTests) > 0 {
		os.Exit(1)
	}
}

File: test_runner/test_runner.js
class TestRunner {
    constructor() {
        this.failedTests = [];
    }

    runTest(test, name) {
        process.stdout.write(`${name}...`);

        try {
            test();
            process.stdout.write(" pass\n");
        } catch (err) {
            this.failedTests.push(`${name}: ${err.message || err}`);
            process.stdout.write(" fail\n");
        }
    }
}

function main() {
    const runner = new TestRunner();

    /* replace: tests */

    if (runner.failedTests.length > 0) {
        runner.failedTests.forEach(failure => {
            console.log(failure);
        });

        process.exit(1);
    }
}

main()
